url29ecd5d728_geodb_db3dcore_tgz-pJ8-de_uos_igf_db3d_dbms_model3d_SegmentNetBuilderJ8/src/de/uos/igf/db3d/dbms/geom/LA.java

```diff
@@ -0,0 +1,1062 @@
+					--i;
+					resultrc += ar[i] * b[i][c];
+				}
+				result[r][c] = resultrc;
+			}
+		}
+		return result;
+	}
+
+	public static double[] mul(double[][] a, double[] x) {
+
+		// { x11 ,
+		// x21 ,
+		// X31 }
+		// { { a11 , a12 , a13 },
+		// { a21 , a22 , a23 } }
+		//
+		double[] result = new double[a.length];
+		for (int r = 0; r < result.length; ++r) {
+			result[r] = dot(a[r], x);
+		}
+		return result;
+	}
+
+	public static double[] mul(double[] x, double[][] a) {
+
+		// { { a11 , a12 , a13 },
+		// { a21 , a22 , a23 } }
+		// { x1, x2 }
+		//
+		double[] result = new double[0];
+		for (int r = 0; r < x.length; ++r) {
+			result = plus(result, mul(x[r], a[r]));
+		}
+		return result;
+	}
+
+	public static BigDecimal[][] mul(BigDecimal[][] a, BigDecimal[][] b) {
+
+		// { { b11 , b12 , b13 , b14 } ,
+		// { b21 , b22 , b23 , b24 } ,
+		// { b31 , b32 , b33 , b34 } }
+		// { { a11 , a12 , a13 },
+		// { a21 , a22 , a23 } }
+		//
+
+		int width = 0;
+		for (BigDecimal[] bRow : b) {
+			int bRowLength = bRow.length;
+			if (bRowLength > width) {
+				width = bRowLength;
+			}
+		}
+		BigDecimal[][] result = new BigDecimal[a.length][width];
+		for (int r = 0; r < result.length; ++r) {
+			for (int c = 0; c < width; ++c) {
+				BigDecimal resultrc = BigDecimal.ZERO;
+				if (a[r] == null) {
+					continue;
+				}
+				BigDecimal[] ar = a[r];
+				for (int i = Math.min(ar.length, b.length); i > 0;) {
+					--i;
+					resultrc = resultrc.add(ar[i].multiply(b[i][c]));
+				}
+				result[r][c] = resultrc;
+			}
+		}
+		return result;
+	}
+
+	public static double[] solve(int[] piRowCols, double[][] M, double[] y)
+			throws Exception {
+		return solve(piRowCols, piRowCols, M, y);
+	}
+
+	public static double[] solveSym(int[] piRowCols, double[][] M, double[] y)
+			throws Exception {
+		return solveSym(piRowCols, piRowCols, M, y);
+	}
+
+	/**
+	 * Using the LAPACK solver.
+	 * 
+	 * @param M
+	 *            the matrix of the equations system
+	 * @param y
+	 *            the inhomogenious part of the system
+	 * @return an array x of doubles such that Mx = y;
+	 * @throws Exception
+	 *             if the equation system doesn't have exactly one such
+	 *             solution.
+	 */
+	public static double[] solve(int[] piRows, int[] piCols, double[][] M,
+			double[] y) throws Exception {
+
+		int mRows = piRows.length;
+		int mCols = piCols.length;
+
+		if (mRows != mCols) {
+			throw new Exception(Db3dSimpleResourceBundle
+					.getString("db3d.geom.notsqmat"));
+		}
+
+		double[][] pM = new double[mRows][mCols + 1];
+		double[] py = new double[mRows];
+
+		for (int c = 0; c < mCols; ++c) {
+			try {
+				py[c] = y[piCols[c]];
+			} catch (ArrayIndexOutOfBoundsException ex) {
+				py[c] = 0d;
+			}
+		}
+
+		for (int r = 0; r < mRows; ++r) {
+			for (int c = 0; c < mCols; ++c) {
+				try {
+					pM[r][c] = M[piRows[r]][piCols[c]];
+				} catch (ArrayIndexOutOfBoundsException ex) {
+					pM[r][c] = 0d;
+				}
+			}
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 975:983, 1023:1031.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.